### ISSUE TITLE
Improve gammapy.datasets

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#288]
The [gammapy.datasets](https://gammapy.readthedocs.org/en/latest/datasets/index.html) sub-package should give easy access to commonly used datasets for Gammapy users and for high-level tests.

IMO the current implementation and how we use it has several serious issues:
- Files are downloaded to the Astropy cache (on my machine located in `~/.astropy/cache/download/`) and stored in files where the filename is a sha such as `1259686d4790c86a63b426698acceebc`. This makes it hard for users to find a file directly, e.g. to open an image in ds9. It's also not a good solution for big files, because users might get annoyed that their hard disk fills up in a hidden `.astropy` cache folder. Also in most functions there is no `cache` option or `cache=True` is the default, i.e. users might get files from their cache and produce outdated results without realising it.
- Some of the dataset access functions return Gammapy objects defined in other sub-packages and some of the examples in other sub-packages use dataset access functions. This has lead to circular import issues in the past, as noted in #171. As an example, [gammapy.datasets.FermiGalacticCenter.diffuse_model](https://gammapy.readthedocs.org/en/latest/api/gammapy.datasets.FermiGalacticCenter.html#gammapy.datasets.FermiGalacticCenter.diffuse_model) returns a `gammapy.data.SpectralCube` object. The other way around, we want to load EBL absorption model tables or other data files to implement some functionality in other Gammapy sub-packages, i.e. import `gammapy.datasets` from there.
- There are quite a few tests that load test data (either bundled or remote) via `gammapy.datasets`. It is currently not possible to read the remote datasets from the cache, i.e. they have to be re-fetched every time a test is run.
- Some file URLs on the web aren't stable enough (see e.g. #287). That will always be an issue with some files / services. But it could make sense to store our own copy (e.g. in `gammapy-extra`) or have a `gammapy-datasets.json` index file in the `gammapy-extra` github repo that redirects to the latest URL.
- Some of the datasets we want to download will be big (GB to TB size), e.g. all Fermi or HESS (see [DataStore](https://gammapy.readthedocs.org/en/latest/api/gammapy.obs.DataStore.html) or CTA science challenge data. At least for those use-cases, but actually in general, it would be best if we had a way to compute the file size we'll download and print it to the console and also have a progress bar.

Now what needs to be done?

Some first thoughts ... other opinions and suggestions very welcome!
- Look closely what others have done: `scikit-learn`, `astroml`, [skdata](https://github.com/jaberg/skdata), [bokeh](http://bokeh.pydata.org/en/latest/docs/installation.html#sample-data), [astro data](https://astropy.stsci.edu/data/)
- Don't use the Astropy cache, instead have one (or a few?) configurable download directory where files are stored with the filenames they have on the web. It shouldn't be too hard to come up with sensible folder names to structure this locally. Example: https://astropy.stsci.edu/data/
- Have a separate list of data that's only useful for tests and not of interest to end-users, not even for examples. A lot of tests need input data and faking it is often a lot of work. So I'm thinking it should be OK for someone running tests locally or for travis-ci to _always_ have to download ~10 MB or at least < 100 MB of test data files. Example: https://github.com/sherpa/sherpa-test-data
- There should be a utility function like [here](http://bokeh.pydata.org/en/latest/docs/installation.html#sample-data) that downloads the majority of data (say < 100 MB). Maybe with an option to opt-in to download big datasets like all Fermi or HESS data, or just keep those separate.
- Maybe get rid of all the `load_` functions that return Gammapy objects. In tutorials and examples I think this is probably not a good way anyways, because it doesn't show users how they would load _their_ equivalent data file. So this is bad:

``` python
from gammapy.datasets import load_something
something = load_something()
```

and this would be better even if a bit longer?

``` python
from gammapy.datasets import get_path
filename = get_path('abc/something.fits')
from gammapy.abc import SomeThing
something = SomeThing.read(filename)
```

This would get rid of the circular import issue because `gammapy.datasets` would not import other sub-packages.

**Comment by cdeil** [comment]
I recently discussed this with @adonath and we settled on a setup where `gammapy-extra` is a git submodule. This is done for `ctapipe` and even documented a little:
https://github.com/cta-observatory/ctapipe/blob/master/docs/datasets/index.rst
The same should be done for Gammapy.

**Comment by cdeil** [comment]
Another good example for how to configure data download:
http://docs.sunpy.org/en/stable/guide/acquiring_data/index.html#acquiring-data-with-sunpy
Also, their database might be a useful model for out `Datastore`.

**Comment by cdeil** [comment]
This was partly implemented in #364.

**Comment by cdeil** [comment]
Closing this old issue now.
